Validate target port

nickandreev opened the pull request:
# Description

Fixes #128 

Check if target port is used by all target pods.


# Checklist:

- [x] My code follows the coding style of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works.   
- [ ] Any dependent changes have been merged and published in downstream modules<br>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/xk6-disruptor?pullRequest=136) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/xk6-disruptor?pullRequest=136) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/xk6-disruptor?pullRequest=136) it.</sub>

pablochacin reviewed (changes requested):
Thanks for the PR @nickandreev. I have some feedback that I think simplifies the code.

The most important one is regarding the need of adding the target port to the test cases instead of changing the target port in the fault injection. I think that way we can test the same logic but without adding more complexity to the test cases.

nickandreev commented on pkg/disruptors/pod.go:
```diff
@@ -104,13 +108,61 @@ func (d *podDisruptor) Targets() ([]string, error) {
+// validateTargetPort checks if every target pod listens to target port
+func (d *podDisruptor) validateTargetPort(targetPort uint) error {
+	if targetPort == 0 {
+		targetPort = protocol.DefaultTargetPort
```
I agree that having two defaults is like a parallel API. I didn't understand the idea behind default 80 on the agent side and I also think that it would be better to make it a required parameter 👍  
I will create a follow-up issue

nickandreev commented on pkg/disruptors/pod.go:
```diff
@@ -104,13 +108,61 @@ func (d *podDisruptor) Targets() ([]string, error) {
+		return err
+	}
+
+	for _, target := range targets {
+		pod, err := d.k8s.CoreV1().Pods(d.selector.Namespace).Get(d.ctx, target, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		found := false
+		for _, container := range pod.Spec.Containers {
+			for _, containerPort := range container.Ports {
+				if uint(containerPort.ContainerPort) == targetPort {
+					found = true
+					break
+				}
+			}
+			if found {
+				break
+			}
+		}
+		if !found {
+			return fmt.Errorf("target %q doesn't listen to port %d", target, targetPort)
+		}
+	}
```
That looks cleaner, thank you 🙏

pablochacin commented on pkg/disruptors/pod_test.go:
```diff
@@ -334,7 +432,25 @@ func Test_PodGrpcPFaultInjection(t *testing.T) {
+					port = 80
+				}
+				obj := builders.NewPodBuilder(target).
+					WithContainerPort(port).
```
I see, but I think this introduces too much complexity just for adding this test case. Also, adding a parameter to the test case struct for only one case generally smells like trying to test some logic from the wrong perspective.

I had the same issue when I tried to test the service port mapping in this [PR]( https://github.com/grafana/xk6-disruptor/pull/131), so what I did there was to refactor that code in a separate structure, the port mapper and then test it independently. 

I think you could follow a similar path here. I don't have a strong opinion now about how this should be done and maybe this is a little overkill here to introduce a `PortValidator` or something like that. 

Therefore, even when I'm not a fan of testing internal functions in unit tests, if I have to choose between that or adding this complexity to the tests, I would prefer to create a new unit test for the `findPort` function and we can reconsider refactoring it in a follow-up PR.

nickandreev replied:
I've tried a separate `PortValidator` and it looks much better 👍

pablochacin replied:
Looks great.

nickandreev commented:
Thank you @pablochacin  for your comments! I pushed new commit that includes refactored validation logic and new tests

pablochacin commented on pkg/disruptors/pod_test.go:
```diff
@@ -334,7 +394,25 @@ func Test_PodGrpcPFaultInjection(t *testing.T) {
+				obj := builders.NewPodBuilder(target).
+					WithLabels(tc.selector.Select.Labels).
+					WithNamespace(tc.selector.Namespace).
+					WithContainerPort(DefaultTargetPort).
```
I think this is not necessary, is it? I find it confusing we are setting this in this test.

nickandreev replied:
You mean container port? It's needed in order to prevent all existing test cases from failing. I can replace `PortValidator` with a dummy one with empty targets array. In this case I can avoid k8s objects mocking entirely but I won't be able to test the case when validator throws an error. Another option would be introducing an interface for PortValidator and have `fakePortValidator` for disruptor tests (as it is for AgentController). This approach will make disruptor tests cleaner and easier to understand  but it will introduce new interface in code.

BTW, I noticed that I forgot to remove `containerPort` in test case struct and I left `expectedCmd` in new tests, I will clean that 👌

pablochacin replied:
Ok, I see. I would prefer that the `PodBuilder` provided that default container port instead of requiring to add it on each test, so that `NewPodBuilder(..).Build()` should provide a default pod that can be used in most test cases.

nickandreev replied:
Oh, that's indeed better, I overcomplicated it 😅

nickandreev replied:
And what do you think about getting rid of mocking k8s objects and mocking PortValidator instead?

nickandreev replied:
Fixed. But now I noticed that there is an implicit dependency on default port value between builders (both pod and service) and disruptors packages. It doesn't look harmful now but maybe it's worth refactoring in future

pablochacin replied:
Hi @nickandreev do you think the PR is ready from your side? I plan to make a final review and merge it.

pablochacin commented on e2e/disruptors/pod_e2e_test.go:
Nitpicking: When possible, changes not related to the PR like fixing typos or lining code should be in a separate commit to facilitate reviewers to focus on the significant changes.

nickandreev commented on pkg/disruptors/pod.go:
```diff
@@ -104,13 +111,79 @@ func (d *podDisruptor) Targets() ([]string, error) {
+		targetPort = DefaultTargetPort
+	}
+
+	for _, target := range pv.targets {
```
Hi @pablochacin ! I think it's almost ready. The last thing that concerns me is that there is a potential performance issue: currently pods are retrieved sequentially in PortValidator. If disruptor will be launched on, let's say, 50 pods, there may be a significant delay. I can parallelise these sequential check with goroutines but the logic probably will be too complex for such trivial task. We can merge it and I'll create a follow-up issue for this or I can try to figure out other options how to overcome this.

pablochacin replied:
Yes, I have also noticed this performance issue. I think we better consider options before merging this, because as you said it can have a significant performance penalty.

pablochacin replied:
Hi @nickandreev I've been thinking about alternatives to the performance issue. I think the problem comes from making the validation at the wrong place, where the information about the pods is no longer available (only the names).  

The `PodSelector`  has the [logic for retrieving the list of pods in one call](https://github.com/grafana/xk6-disruptor/blob/main/pkg/disruptors/selector.go#L47-L74), but it is not clear to me is how to reuse this logic without breaking encapsulation or creating leaking abstractions. 

Talking about leaking abstractions, this is an issue I have identified in another place: currently the `agentContoller` is [checking if the agent is injected](https://github.com/grafana/xk6-disruptor/blob/main/pkg/disruptors/controller.go#L73-L85) before injecting it. This requires retrieving the Pod using the K8S client and calling the `PodHelper` to inject the agent.  I don't like this code because it breaks the encapsulation of the Kubernetes-related code into the `kubernetes` package and creates some issues for testing the disruptor package.

My intention is to address these issues by [refactoring the Kubernetes-related logic into the `Kubernetes` package](https://github.com/grafana/xk6-disruptor/issues/139).

Maybe after this work is done (hopefully this same week) we can revisit the implementation of the validation.

pablochacin reviewed (changes requested):
As commented [here](https://github.com/grafana/xk6-disruptor/pull/136#discussion_r1173568056)  postpone merging this change until the potential performance penalty is addressed.

pablochacin commented:
Hi @nickandreev I finished [refactoring the Kubernetes-related code](https://github.com/grafana/xk6-disruptor/pull/143). You will notice there are changes that impact your PR. The most important is that all Kubernetes-related functionality was moved to the [Kubernetes helpers package](https://github.com/grafana/xk6-disruptor/tree/main/pkg/kubernetes/helpers). 

Therefore the pod port validation you are implementing should be moved there.

I'm sorry that this impacts your work in progress, but I'm confident the new code structure is clearer and easier to work with.

nickandreev commented:
Hi @pablochacin ! Sorry for the delayed reply, I will check out your changes and try to adjust my PR 👍

pablochacin commented on pkg/disruptors/pod.go:
```diff
@@ -120,6 +130,37 @@ func (d *podDisruptor) InjectGrpcFaults(
+		port = DefaultTargetPort
+	}
+
+	return pv.helper.ValidatePort(ctx, pv.filter, port)
```
The PortValidator is basically just calling the helper function. Given that the pod disruptor already uses this helper I think we don't really need the validator. If I recall correctly, we introduced it just to facilitate testing. But as the validation logic is now a function of the Helper, we can test it there.

nickandreev replied:
Agree, for some reason I decided to avoid direct dependencies in PodDisruptor, but now I see it doesn't make sense

pablochacin commented on pkg/kubernetes/helpers/pods_test.go:
```diff
@@ -440,3 +440,52 @@ func Test_ListPods(t *testing.T) {
 	}
 }
+
+func Test_VerifyPort(t *testing.T) {
```
nit: maybe use the same name of the function you are testing, `ValidatePort`

pablochacin commented on pkg/disruptors/pod_test.go:
```diff
@@ -358,3 +428,78 @@ func Test_PodGrpcPFaultInjection(t *testing.T) {
+		expectError bool
+	}{
+		{
+			title:     "Successful port verification",
```
If we remove the `PortValidator` I think these test cases should be moved to the pod helper tests.

pablochacin commented on pkg/testutils/kubernetes/builders/pod.go:
```diff
@@ -57,6 +57,12 @@ func (b *podBuilder) WithContainer(c corev1.Container) PodBuilder {
 func (b *podBuilder) Build() *corev1.Pod {
+	containers := b.containers
+
+	if containers == nil {
+		containers = defaultContainers()
+	}
```
This function is used in many tests. I don't think ti should be too opinionated as to defined some default containers.
If you want to simply the setup of some specific tests (for instance, the pod helper tests), create a build container function there.

pablochacin commented on pkg/testutils/kubernetes/builders/pod.go:
```diff
@@ -68,11 +74,26 @@ func (b *podBuilder) Build() *corev1.Pod {
 	}
 }
+
+func defaultContainers() []corev1.Container {
```
I would move this function to the test that needs it.

pablochacin commented:
Hi @nickandreev thanks for continuing with the PR. I think we are closer now  :crossed_fingers:
I made the comments in line. 

My only significant concern is that I think we don't need the port validator anymore, after the refactoring introduced in #143. 

If I recall correctly, it was introduced then to facilitate testing. But now, it basically has no logic and its test duplicate those of the `ValidatePort` function in the `PodHelper`.

I suggest removing the `PortValidator` and calling the `ValidatePort` function directly from the PodDisruptor. Also, I think we can move any tests that are not duplicated to the `PodHelper`

nickandreev commented:
Hi @pablochacin ! Thanks for your comments and sorry for such a slow PR 😅 
I pushed updated code, please, have a look when you have time 👍

pablochacin commented on pkg/disruptors/pod.go:
```diff
@@ -108,7 +115,12 @@ func (d *podDisruptor) InjectHTTPFaults(
-	err := d.controller.ExecCommand(ctx, cmd)
+	err := d.validatePort(ctx, fault.Port)
+	if err != nil {
+		return fmt.Errorf("verify fault port: %w", err)
```
I think this error description could be reworded as "port validation error: " or something similar for clarity

pablochacin commented on pkg/disruptors/pod.go:
```diff
@@ -120,6 +132,20 @@ func (d *podDisruptor) InjectGrpcFaults(
+
+	err := d.validatePort(ctx, fault.Port)
+	if err != nil {
+		return fmt.Errorf("verify fault port: %w", err)
```
Same as before

pablochacin reviewed (changes requested):
Hi @nickandreev

Thanks for your updates to the PR. It looks good to me except for two minor issues I mentioned in the code.

However, the biggest issue I have before we can consider merging it is that you have been merging the `main` branch into your branch instead of rebasing your branch. That is creating a very weird git history and I'm not sure how it will look like when merging your branch into main.

nickandreev commented:
Hi @pablochacin , sorry for the mess in commit history. I fixed the issues you mentioned and squashed all commits. Also, I fixed the test for PodHelper, it missed one test case 👍

pablochacin reviewed (approved):
LGTM

Thanks, @nickandreev for your commitment to complete this PR.

pablochacin commented:
oops, something is broken in the unit tests. That's weird because the previous version (before the squash in one commit) were passing

```
--- FAIL: Test_JsPodDisruptor (0.01s)
    --- FAIL: Test_JsPodDisruptor/inject_HTTP_Fault_with_full_arguments (0.01s)
        api_test.go:411: failed GoError: error injecting fault: validate fault port: no pods were selected by specified filter at reflect.methodValueCall (native)
    --- FAIL: Test_JsPodDisruptor/inject_Grpc_Fault_without_options (0.01s)
        api_test.go:411: failed GoError: error injecting fault: validate fault port: no pods were selected by specified filter at reflect.methodValueCall (native)
    --- FAIL: Test_JsPodDisruptor/inject_HTTP_Fault_without_options (0.01s)
        api_test.go:411: failed GoError: error injecting fault: validate fault port: no pods were selected by specified filter at reflect.methodValueCall (native)
    --- FAIL: Test_JsPodDisruptor/inject_Grpc_Fault_with_full_arguments (0.01s)
        api_test.go:411: failed GoError: error injecting fault: validate fault port: no pods were selected by specified filter at reflect.methodValueCall (native)
FAIL
````

pablochacin reviewed (changes requested):
oops, something is broken in the unit tests:

<details>

<summary>

```
--- FAIL: Test_JsPodDisruptor (0.01s)
    --- FAIL: Test_JsPodDisruptor/inject_HTTP_Fault_with_full_arguments (0.01s)
```

</summary>

```
        api_test.go:411: failed GoError: error injecting fault: validate fault port: no pods were selected by specified filter at reflect.methodValueCall (native)
    --- FAIL: Test_JsPodDisruptor/inject_Grpc_Fault_without_options (0.01s)
        api_test.go:411: failed GoError: error injecting fault: validate fault port: no pods were selected by specified filter at reflect.methodValueCall (native)
    --- FAIL: Test_JsPodDisruptor/inject_HTTP_Fault_without_options (0.01s)
        api_test.go:411: failed GoError: error injecting fault: validate fault port: no pods were selected by specified filter at reflect.methodValueCall (native)
    --- FAIL: Test_JsPodDisruptor/inject_Grpc_Fault_with_full_arguments (0.01s)
        api_test.go:411: failed GoError: error injecting fault: validate fault port: no pods were selected by specified filter at reflect.methodValueCall (native)
FAIL
````

</details>


The issue is that the validation we are doing is throwing an error if there are no pods matching the selector. I think we should only throw an error if there are pods and at least one of them does not listen to the target port.

nickandreev commented:
removed that condition, now everything is green 🙂

pablochacin reviewed (approved):
LGTM

nickandreev commented:
Thank you @pablochacin for you support and help! 🙏
